src/shared/psi-util.c

```diff
@@ -0,0 +1,115 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <stdio.h>
+#include <unistd.h>
+
+#include "alloc-util.h"
+#include "extract-word.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "parse-util.h"
+#include "psi-util.h"
+#include "string-util.h"
+#include "stat-util.h"
+#include "strv.h"
+
+int read_resource_pressure(const char *path, PressureType type, ResourcePressure *ret) {
+        _cleanup_free_ char *line = NULL, *word = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        ResourcePressure rp = {};
+        const char *t, *cline;
+        bool field_filled[] = { false, false, false, false };
+        int r;
+
+        assert(path);
+        assert(IN_SET(type, PRESSURE_TYPE_SOME, PRESSURE_TYPE_FULL));
+        assert(ret);
+
+        if (type == PRESSURE_TYPE_SOME)
+                t = "some";
+        else if (type == PRESSURE_TYPE_FULL)
+                t = "full";
+        else
+                return -EINVAL;
+
+        f = fopen(path, "re");
+        if (!f)
+                return -errno;
+
+        for (;;) {
+                _cleanup_free_ char *l = NULL;
+                char *w;
+
+                r = read_line(f, LONG_LINE_MAX, &l);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                w = first_word(l, t);
+                if (w) {
+                        line = TAKE_PTR(l);
+                        break;
+                }
+        }
+
+        if (!line)
+                return -ENODATA;
+
+        /* skip type string */
+        cline = line + 4;
+
+        /* extracts either avgX=Y.Z or total=X */
+        while ((r = extract_first_word(&cline, &word, NULL, 0)) > 0) {
+                _cleanup_free_ char *a = NULL, *b = NULL;
+                const char *cword;
+
+                /* extracts (avgX and Y.Z) or (total and X) */
+                cword = word;
+                r = extract_many_words(&cword, "=", 0, &a, &b, NULL);
+                word = mfree(word);
+                if (r < 0)
+                        return r;
+                if (r != 2)
+                        return -EINVAL;
+
+                if (streq(a, "avg10")) {
+                        field_filled[0] = true;
+                        r = parse_loadavg_fixed_point(b, &rp.avg10);
+                } else if (streq(a, "avg60")) {
+                        field_filled[1] = true;
+                        r = parse_loadavg_fixed_point(b, &rp.avg60);
+                } else if (streq(a, "avg300")) {
+                        field_filled[2] = true;
+                        r = parse_loadavg_fixed_point(b, &rp.avg300);
+                } else if (streq(a, "total")) {
+                        field_filled[3] = true;
+                        r = parse_time(b, &rp.total, 1);
```

Comment:
parse_time() is for "human readable" time, i.e. with units such as "s" or "h" or so. this is not like that is it? if so, just use plain safe_atou64()